test/distributed/fsdp/test_fsdp_optim_state.py

```diff
@@ -0,0 +1,435 @@
+        for full_osd_state in full_osd_states:
+            # Check for at least one match (may be > 1 in toy edge cases, e.g.
+            # multiple biases); nonetheless, each having >= 1 match and the two
+            # lists have equal length imply that the list contents are equal
+            self.assertTrue(any(
+                self._are_equal_states(full_osd_state, ref_osd_state)
+                for ref_osd_state in ref_osd_states
+            ))
+
+    def _check_same_param_groups(self, full_osd, ref_osd):
+        """Checks that ``full_osd`` and ``ref_osd`` have the same
+        "param_groups" part."""
+        assert "param_groups" in ref_osd
+        self.assertTrue("param_groups" in full_osd)
+        ref_osd_param_groups = ref_osd["param_groups"]
+        full_osd_param_groups = full_osd["param_groups"]
+        self.assertTrue(len(full_osd_param_groups), len(ref_osd_param_groups))
+        for full_osd_pg, ref_osd_pg in zip(full_osd_param_groups, ref_osd_param_groups):
+            self.assertEqual(set(full_osd_pg.keys()), set(ref_osd_pg.keys()))
+            for name, full_osd_value in full_osd_pg.items():
+                # Even if the parameter IDs map differently to parameters,
+                # "params" should still contain the same IDs and be in
+                # increasing order; thus, the two values should equal
+                self.assertEqual(full_osd_value, ref_osd_pg[name])
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("use_multiple_param_groups", [False, True])
+    def test_full_optim_state_dict_affine(
+        self,
+        use_multiple_param_groups: bool,
+    ) -> None:
+        """
+        Tests :meth:`full_optim_state_dict` by comparing the returned dict for
+        an FSDP-wrapped model with that of an equivalent non-wrapped model.
+
+        The parameter groups in the "param_groups" part and the values in the
+        "state" part should be the same, but the parameter IDs (i.e. the keys
+        in the "state" part) may be rearranged.
+        """
+        NUM_ITERS = 3
+        model1, optim1, optim_input = self._init_affine_model(
+            wrap=True, use_multiple_param_groups=use_multiple_param_groups,
+        )
+        losses1 = self._step_model(model1, optim1, num_iters=NUM_ITERS)
+        full_osd = FSDP.full_optim_state_dict(model1, optim1, optim_input)
+        if self.rank != OPTIM_TARGET_RANK:
+            return
+
+        model2, optim2, _ = self._init_affine_model(
+            wrap=False, use_multiple_param_groups=use_multiple_param_groups,
+        )
+        losses2 = self._step_model(model2, optim2, num_iters=NUM_ITERS)
+        ref_osd = optim2.state_dict()
+
+        for i, (l1, l2) in enumerate(zip(losses1, losses2)):
+            assert l1 == l2, f"Losses differ on iter {i}: {l1:.5f} {l2:.5f}"
+
+        self._check_same_param_groups(full_osd, ref_osd)
+        self._check_same_state(full_osd, ref_osd, check_same_param_ids=False)
+
+    @skip_if_lt_x_gpu(2)
```

Comment:
These tests for `shard_full_optim_state_dict()` are more meaningful when the test environment has 4 GPUs. Otherwise, the halved-world-size model does not shard parameters since its world size is 1.